src/Data/Text/Utf8.hs

```diff
@@ -214,21 +235,43 @@ decodeUtf8 cus = error $ "Invalid UTF-8 input sequence at " ++ show (take 4 cus)
 -- | Does exactly the same thing as 'unsafeIndexCodePoint'', but on 'Text' values.
 {-# INLINE unsafeIndexCodePoint #-}
 unsafeIndexCodePoint :: Text -> CodeUnitIndex -> (CodeUnitIndex, CodePoint)
-unsafeIndexCodePoint (Text !u8data !off !_len) (CodeUnitIndex !index) =
-  unsafeIndexCodePoint' u8data $ CodeUnitIndex $ off + index
+unsafeIndexCodePoint (Text !u8data !off !_len) !index =
+  unsafeIndexCodePoint' u8data $ CodeUnitIndex off + index
 
 -- | Get the code unit at the given 'CodeUnitIndex'.
 -- Performs bounds checking.
 {-# INLINE indexCodeUnit #-}
 indexCodeUnit :: Text -> CodeUnitIndex -> CodeUnit
-indexCodeUnit !text (CodeUnitIndex !index)
-  | index < 0 || index >= codeUnitIndex (lengthUtf8 text) = error $ "Index out of bounds " ++ show index
-  | otherwise = unsafeIndexCodeUnit text $ CodeUnitIndex index
+indexCodeUnit !text !index
+  | index < 0 || index >= lengthUtf8 text = error $ "Index out of bounds " ++ show index
+  | otherwise = unsafeIndexCodeUnit text index
 
 {-# INLINE unsafeIndexCodeUnit #-}
 unsafeIndexCodeUnit :: Text -> CodeUnitIndex -> CodeUnit
-unsafeIndexCodeUnit (Text !u8data !off !_len) (CodeUnitIndex !index) =
-  unsafeIndexCodeUnit' u8data $ CodeUnitIndex $ off + index
+unsafeIndexCodeUnit (Text !u8data !off !_len) !index =
+  unsafeIndexCodeUnit' u8data $ CodeUnitIndex off + index
+
+-- | Scan backwards through the text until we've seen the specified number of codepoints. Assumes
+-- that the initial CodeUnitIndex is within a codepoint.
+{-# INLINE skipCodePointsBackwards #-}
+skipCodePointsBackwards :: Text -> CodeUnitIndex -> Int -> CodeUnitIndex
+skipCodePointsBackwards (Text !u8data !off !len) !index0 !n0
+  | index0 >= CodeUnitIndex len = error "Invalid use of skipCodePointsBackwards"
+  | otherwise = loop (index0 + CodeUnitIndex off) n0
+  where
+    loop index n | atTrailingByte index =
+      loop (index-1) n  -- Don't exit before we're at a leading byte
+    loop index 0 | index < 0 =
+      -- We don't prevent reads outside the array, but you do get an error afterwards
```

Comment:
Ah, maybe the comment just had me confused. Do we expect it to read out of bounds during regular operations, or is this purely a safeguard against bugs? Because it read like it would by design read out of bounds first, only to then figure out that it shouldn't have done that.

Reply:
Aaah ok, yes it's purely a safeguard against bugs. Does this phrasing make it a bit clearer?
```haskell
      -- Throw an error if we've read before the array, this one-time check
      -- doesn't prevent segfaults but may help you locate bugs.
```